pkg/collectsub/datasource/csubsource/csubsource.go

```diff
@@ -0,0 +1,113 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package csubsource
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/guacsec/guac/pkg/collectsub/client"
+	pb "github.com/guacsec/guac/pkg/collectsub/collectsub"
+	"github.com/guacsec/guac/pkg/collectsub/datasource"
+)
+
+type csubDataSources struct {
+	c            client.Client
+	lastEntries  *datasource.DataSources
+	pollDuration time.Duration
+}
+
+// NewFileDataSources creates a datasource which gets its data sources
+// from a configuration file. This configuration file is in YAML and
+// follows the structure outlined in the FileFormat struct. An example
+// is as follows:
+func NewCsubDatasource(c client.Client, pollDuration time.Duration) (datasource.CollectSource, error) {
+	return &csubDataSources{
+		c:            c,
+		pollDuration: pollDuration,
+	}, nil
+}
+
+// GetDataSources returns a data source containing targets for the
+// collector to collect
+func (d *csubDataSources) GetDataSources(ctx context.Context) (*datasource.DataSources, error) {
+	entries, err := d.c.GetCollectEntries(ctx, []*pb.CollectEntryFilter{
+		{Type: pb.CollectDataType_DATATYPE_OCI, Glob: "*"},
+		{Type: pb.CollectDataType_DATATYPE_GIT, Glob: "*"},
+	})
+	if err != nil {
+		return nil, err
+	}
+	ds := entriesToSources(entries)
+	d.lastEntries = ds
+
+	return ds, nil
+}
+
+// DataSourcesUpdate will return a channel which will get an element
+// if the CollectSource has new data. Upon update, nil is inserted
+// into the channel and non-nil if the channel no longer is able to
+// serve updates.
+func (d *csubDataSources) DataSourcesUpdate(ctx context.Context) (<-chan error, error) {
+	updateChan := make(chan error)
+	go func() {
+		for {
+			select {
+			case <-time.After(d.pollDuration):
```

Comment:
I would recommend using `ticker := time.NewTicker(d.pollDuration)`
The use of a ticker instead of time.After can improve the reliability of the code in several ways:

1. A ticker is a periodic timer that repeatedly sends a value on a channel, while time.After sends a value only once after a specified duration. This means that a ticker can be used to run a task at regular intervals, while time.After would only trigger once after the specified duration.

1. The ticker's channel can be stopped or reset, providing more control over the timing of the periodic task. With time.After, once the specified duration has passed, there is no way to stop or reset the timer.

1. Tickers are more efficient than using time.After in a loop, as they use a single underlying timer. Each time time.After is used, a new timer is created, leading to increased resource usage.

In short, tickers provide a more flexible and efficient way to perform periodic tasks compared to using time.After.

Reply:
hmm yea that seems suitable! I'm guessing probably a ticker + do a reset after each select is completed so that the ticker doesn't get backlogged if the logic takes too long. Learn something new everyday. Thanks!

Reply:
👍